stdlib/asm/crypto/dsa/falcon.masm

```diff
@@ -1,278 +1,611 @@
+
+    # Compute pi1(tau)
+    repeat.64
+        mem_stream
+        repeat.8
+            exec.deep_queries::combine_main
+        end
+    end
+    #=> [X, X, ev1, ev0, ev1, ev0, pi_ptr, zeros_ptr, tau_ptr, 0, ...]
+    # where (ev0, ev1) := pi1(tau)
+    
+    # Save pi_1(tau)
+    swapw.2 loc_storew.3
+
+    # Setup the pointers
+    swapw.3 loc_loadw.0 add.384
+
+    # Accumulator to compute pi2(tau)
+    swapw dropw padw
+    swapdw
+
+    # Compute pi2(tau)
+    repeat.64
+        mem_stream
+        repeat.8
+            exec.deep_queries::combine_main
+        end
+    end
+    #=> [X, X, ev1, ev0, ev1, ev0, pi_ptr, zeros_ptr, tau_ptr, 0, ...]
+
+    # Load tau^512
+    swapw.3
+    #=> [pi_ptr, zeros_ptr, tau_ptr, 0, X, ev1, ev0, ev1, ev0, X, ...]
+    dup.2 mem_loadw drop drop
+    #=> [tn1, tn0, X, ev1, ev0, ev1, ev0, X, ...]
+
+    # Compute (a0, a1) := tau^512 * pi2(tau)
+    movup.7 movup.7 ext2mul
+    #=> [a1, a0, X, x, x, X, ...]
+
+    # Compute (res0, res1) := pi(tau)
+    swapw loc_loadw.3 drop drop
+    ext2add
+    #=> [res1, res0, x, x, X, ...]
+
+
+    # 6) Check product
+
+    ## a) Load h(tau)
+    swapw
+    loc_loadw.1
+
+    ## b) Load s2(tau)
+    push.0.0
+    loc_loadw.2
+
+    ## c) compute the product
+    drop drop
+    ext2mul
+
+    ## d) assert equality
+    movup.2 
+    assert_eq
+    assert_eq
+
+    ## e) clean up the stack
+    drop drop
+    #=> [...]
 end
 
-#! Given four elements on stack top, this routine computes squared norm of that
-#! vector ( read polynomial ) with four coefficients.
-#!
-#! Imagine, given vector is f, which is described as
-#!
-#! f = [a0, a1, a2, a3]
-#!
-#! Norm of that vector is
+
+# SQUARE NORM OF Z_q[x]/(phi) POLYNOMIALS 
+# =============================================================================================
+
+#! Normalizes an `e` in [0, q) to be in [-(q-1) << 1, (q-1) << 1) and returns its square norm.
 #!
-#! √(a0 ^ 2 + a1 ^ 2 + a2 ^ 2 + a3 ^ 2)
+#! We use the following formula to do so:
+#! normalize(e) = e^2 - phi * (2*q*e - q^2) where phi := (e > (q - 1)/2)
 #!
-#! But we need squared norm, which is just skipping the final square root operation.
+#! The formula implements:
 #!
-#! Expected stack state :
+#! if e > (q-1)/2:
+#!   return (q - e)^2
+#! else:
+#!   return e^2 
 #!
-#! [a0, a1, a2, a3, ...]
+#! The use of the formula avoids using the if-else block.
 #!
-#! Final stack state :
+#! Input: [e, ...]
+#! Output [norm(e)^2, ...]
 #!
-#! [b, ...] | b = a0 ^ 2 + a1 ^ 2 + a2 ^ 2 + a3 ^ 2
-proc.squared_norm_word
+#! Cycles: 21
+export.norm_sq
+    
     dup
-    mul
-
+    dup mul
     swap
-    dup
-    mul
+    #=> [e, e^2, ...]
 
-    add
-
-    swap
     dup
-    mul
-
-    add
+    push.6144
+    u32unchecked_gt
+    #=> [phi, e, e^2]
 
     swap
-    dup
-    mul
+    mul.24578       # 2*q 
+    push.151019521  # q^2
+    sub
+    #=> [2*q*e - q^2, phi, e^2
 
-    add
+    mul sub
+    #=> [norm(e)^2, ...]
 end
```

Comment:
I wonder if something like this would be a bit more efficient:

- Compute $f = e > (q - 1) / 2$ - i.e., $f=1$ if $e > (q - 1) / 2$, and $f = 0$ otherwise.
- Compute $x = f \cdot q - e$.
- Return $x^2$.

Reply:
The complication is due to the $-e$.  In the formula $f.q -e$, when $f = 0$, $-e$ should be interpreted as a signed integer but the code tries to use only field arithmetic in order to compute things.
I am actually wondering if it would be more efficient to just use an if-else statement like it is described in the procedure comment.

Reply:
> I am actually wondering if it would be more efficient to just use an if-else statement like it is described in the procedure comment.

I'm not sure about this. It could take a few cycles fewer but at the same time would introduce extra nodes to the program MAST - and overall, maybe be less efficient. So, I'm inlined to keep it branchless for now.


> The complication is due to the $-e$.  In the formula $f.q -e$, when $f = 0$, $-e$ should be interpreted as a signed integer but the code tries to use only field arithmetic in order to compute things.

Ah! Makes sense! How about something like this:

- Compute $f$ such that $f=-1$ if $e > (q - 1) / 2$, and $f = 0$ otherwise.
- Compute $x = - (f \cdot q + e)$.
- Return $x^2$.

Or would this also run into similar issues?

Reply:
I think this runs into the same problem as the first proposal. However, going back to your first proposal, I think that we can make the first proposal work using a wrapping sub. Here is a proposal:
```
export.norm_sq
    dup
    push.6144
    u32unchecked_gt
    #=> [phi, e, ...]

    push.12289 mul swap
    #=> [e, q * phi, ...]

    u32overflowing_sub 
    #=> [overflow, q*phi - e, ...]

    push.4294967296 dup.2 sub
    #=> [e - q*phi, overflow, q*phi - e, ...]

    swap cdrop
    #=> [abs(e - q*phi), ...]

    dup mul
    #=> [(e - q*phi)^2, ...]
end
```
Unfortunately, the cycle count does not improve with this new approach.

Reply:
Makes sense. Let's keep it as is for now - maybe we'll come up with some other optimizations later.